man/mlist.1
.Dd April 30, 2020
.Dt MLIST 1
.Os
.Sh NAME
.Nm mlist
.Nd list and filter messages
.Sh SYNOPSIS
.Nm
.Op Fl DFPRST
.Op Fl X Ar str
.br
.Op Fl dfprst
.Op Fl x Ar str
.br
.Op Fl N | Fl n | Fl C | Fl c
.br
.Op Fl i
.Op Ar dirs\ ...
.Sh DESCRIPTION
.Nm
lists messages in the specified maildir folders
.Ar dirs ,
separated by newlines.
If used non-interactively with no specified folders,
.Nm
reads directory names from the standard input.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl D
Only list messages marked as draft.
.It Fl F
Only list messages marked as flagged.
.It Fl P
Only list messages marked as passed
.Pq resent/forwarded/bounced .
.It Fl R
Only list messages marked as replied-to.
.It Fl S
Only list messages marked as seen.
.It Fl T
Only list messages marked as trashed.
.It Fl X Ar str
Only list messages marked with the characters in
.Ar str .
.It Fl d
Don't list messages marked as draft.
.It Fl f
Don't list messages marked as flagged.
.It Fl p
Don't list messages marked as passed
.Pq resent/forwarded/bounced .
.It Fl r
Don't list messages marked as replied-to.
.It Fl s
Don't list messages marked as seen.
.It Fl t
Don't list messages marked as trashed.
.It Fl x Ar str
Don't list messages marked with the characters in
.Ar str .
.Pq Remember to use uppercase characters.
.It Fl C
Only list messages in
.Pa cur .
.It Fl N
Only list messages in
.Pa new .
.It Fl c
Don't list messages in
.Pa cur .
.It Fl n
Don't list messages in
.Pa new .
.It Fl i
Don't print filenames.
Instead, print a one-line summary for each folder,
showing the number of unseen, flagged and total messages,
along with the folder name.
If any of the message selection flags are used,
the number of matching messages is printed as well.
If two or more folders are specified,
a total for all folders will also be printed at the end.
.El
.Pp
Multiple options are regarded as a conjunction.
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr find 1
.Pp
.Lk https://cr.yp.to/proto/maildir.html "Using maildir format"
.Sh AUTHORS
.An Leah Neukirchen Aq Mt leah@vuxu.org
.Sh LICENSE
.Nm
is in the public domain.
.Pp
To the extent possible under law,
the creator of this work
has waived all copyright and related or
neighboring rights to this work.
.Pp
.Lk http://creativecommons.org/publicdomain/zero/1.0/


NEWS.md
## 0.6 (2020-01-18)

* mfwd: prefix Subject with "Fwd:".
* mscan: add dottime formatting.
* mlist: look at maildir/new too for messages.
* Many bug and portability fixes.

## 0.5.1 (2019-03-03)

* mdeliver: preserve mtime in mrefile
* mdirs: add -0 to separate folders by NUL characters
* Fixes for buffer-overflows, found by fuzzing.
* Fixes for memleaks.

## 0.5 (2019-02-09)

* New tool msearch to wrap several mail indexers.
* New zsh completion _mblaze.
* mnext/mprev were removed (you can call `mless +`/`mless -`).
* The GnuPG tools in contrib/ now use gpg2.
* mshow exits with error if it could not extract all attachments
* mrep: add -noquote to disable quoting the message replied to
* mdeliver: keep permissions of messages
* mcom: aborting the editor is now more like delete than cancel
* mcom: add -send to send directly without editing
* mcom: check if mail is formatted sensibly
* mpick: new flag -v for statistics
* mscan: new flag -v for statistics
* magrep: add -h, which is like -p but doesn't print the file name
* mscan: prioritize displaying trashed mail over other markers
* mpick: fix off-by-one in expression parsing
* Many bug fixes.

## 0.4 (2018-08-15)

* New tool mrefile to move or copy messages.
* contrib/mp7m: add application/pkcs7-mime decoder
* mcom: allow setting arbitrary headers from command line
* mcom: add -body to prepopulate drafts
* mcom: mark drafts as seen after sending
* mcom: update flags after mrep/mbnc/mfwd
* mshow: add -B to decode MIME parameters in broken mails
* magrep: add -l flag
* Many bug fixes.

## 0.3.2 (2018-02-13)

* magrep: add *:REGEX to search in any header
* Fix of a buffer overflow in blaze822_multipart.
* Small bug fixes.
* Many documentation improvements by Larry Hynes.

## 0.3.1 (2018-01-30)

* mless: support $NO_COLOR
* mcolor: support $NO_COLOR
* blaze822.h: ensure PATH_MAX is defined
* Improved documentation.
* Many fixes for address parser.

## 0.3 (2018-01-12)

* New tool mflow to reformat format=flowed plain text mails.
* New tool mbnc to bounce mails (send original mail to someone else).
* mshow filters can output raw text now, e.g. for HTML rendering with colors.
* mrep can quote mail that doesn't have a plain text part.
* mcom runs mmime when deemed necessary.
* mhdr can extract MIME parameters.
* New contrib: mrecode
* New contrib: mraw
* mshow regards non-MIME mails as MIME mails with one part now.
* mshow -F to disable MIME filters.
* mpick supports negations now.
* msed can remove headers depending on their value.
* Improved UTF-8 parsing.
* Improved documentation.
* Numerous bug fixes and portability fixes.

## 0.2 (2017-07-17)

* New sequence syntax `m:+n` for `n` messages after message `m`.
* Threading shortcuts `=`, `_`, `^` for `.=`, `._`, `.^`.
* Sequence related errors are now reported.
* minc and mlist normalize slashes in paths.
* mfwd now generates conforming message/rfc822 parts.
* mthread can add optional folders (e.g. your outbox) to resolve message ids.
* mcom now adds Date: just before sending or cancelling the mail.
* VIOLATIONS.md documents how mblaze works with certain common mistakes.
* Full documentation revamp by Larry Hynes.
* Fix rare crash looking for mail body.
* Numerous small bug and portability fixes.

## 0.1 (2017-06-24)

* Initial release


mlist.c
#define _GNU_SOURCE

#include <sys/stat.h>

#include <dirent.h>
#include <fcntl.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "blaze822.h"
#include "blaze822_priv.h"

/*

PRSTDF
prstdf

-N new
-n not new

-C cur
-c not cur

-m age

-r recursive?

*/

static int8_t flags[255];
static int flagsum;
static int flagset;
static int Nflag;
static int Cflag;
static int iflag;

static long icount;
static long iunseen;
static long iflagged;
static long imatched;

static long tdirs;
static long tunseen;
static long tflagged;
static long tcount;
static long tmatched;

void
list(char *prefix, char *file)
{
	char *f = 0;

	if (flagset || iflag) {
		size_t prefixlen;

		f = strstr(file, ":2,");

		if (!f &&
		    prefix &&
		    (prefixlen = strlen(prefix)) &&
		    prefixlen >= 4 &&
		    strcmp(prefix + prefixlen - 4, "/new") == 0)
			f = ":2,";
	}

	if (flagset) {
		int sum = 0;
		if (!f)
			return;
		icount++;
		tcount++;
		char *g;
		for (g = f + 3; *g; g++)
			if (flags[(unsigned int)*g] == -1)
				return;
			else if (flags[(unsigned int)*g] == 1)
				sum++;
		if (sum != flagsum)
			return;
	}

	if (iflag) {
		if (!f)
			return;
		imatched++;
		tmatched++;
		if (!flagset)
			icount++, tcount++;
		if (!strchr(f, 'S'))
			iunseen++, tunseen++;
		if (strchr(f, 'F'))
			iflagged++, tflagged++;
		return;
	}

	if (prefix) {
		fputs(prefix, stdout);
		putc('/', stdout);
	}
	puts(file);
}

#ifdef __linux__
// faster implementation of readdir using a bigger getdents buffer

#include <sys/syscall.h>

struct linux_dirent64 {
	ino64_t d_ino;           /* 64-bit inode number */
	off64_t d_off;           /* 64-bit offset to next structure */
	unsigned short d_reclen; /* Size of this dirent */
	unsigned char d_type;    /* File type */
	char d_name[];           /* Filename (null-terminated) */
};
#define BUF_SIZE 1024000
char buf[BUF_SIZE];

void
listdir(char *dir)
{
	int fd;
	ssize_t nread;
	ssize_t bpos;
	struct linux_dirent64 *d;

	fd = open(dir, O_RDONLY | O_DIRECTORY);
	if (fd == -1) {
		perror("open");
		return;
	}

	while (1) {
		nread = syscall(SYS_getdents64, fd, buf, BUF_SIZE);
		if (nread == -1) {
			perror("getdents64");
			break;
		}

		if (nread == 0)
			break;

		for (bpos = 0; bpos < nread; bpos += d->d_reclen) {
			d = (struct linux_dirent64 *)(buf + bpos);
			if (!MAIL_DT(d->d_type))
				continue;
			if (d->d_name[0] == '.')
				continue;
			list(dir, d->d_name);
		}
	}

	close(fd);
}
#else
void
listdir(char *dir)
{
	DIR *fd;
	struct dirent *d;

	fd = opendir(dir);
	if (!fd)
		return;
	while ((d = readdir(fd))) {
		if (!MAIL_DT(d->d_type))
			continue;
		if (d->d_name[0] == '.')
			continue;
		list(dir, d->d_name);
	}
	closedir(fd);
}
#endif

void
listarg(char *arg)
{
	squeeze_slash(arg);

	struct stat st;
	if (stat(arg, &st) < 0)
		return;
	if (S_ISDIR(st.st_mode)) {
		char subdir[PATH_MAX];
		struct stat st2;
		int maildir = 0;

		long gcount = icount;
		long gunseen = iunseen;
		long gflagged = iflagged;
		long gmatched = imatched;

		icount = 0;
		iunseen = 0;
		iflagged = 0;

		snprintf(subdir, sizeof subdir, "%s/cur", arg);
		if (stat(subdir, &st2) == 0) {
			maildir = 1;
			if (Cflag >= 0 && Nflag <= 0)
				listdir(subdir);
		}

		snprintf(subdir, sizeof subdir, "%s/new", arg);
		if (stat(subdir, &st2) == 0) {
			maildir = 1;
			if (Nflag >= 0 && Cflag <= 0)
				listdir(subdir);
		}

		if (!maildir)
			listdir(arg);

		if (iflag && (imatched || (maildir && !flagset))) {
			tdirs++;
			if (flagset && imatched)
				printf("%6ld matched  %6ld unseen  %3ld flagged  %6ld msg  %s\n",
				    imatched, iunseen, iflagged, icount, arg);
			else
				printf("%6ld unseen  %3ld flagged  %6ld msg  %s\n",
				    iunseen, iflagged, icount, arg);
		}

		icount = gcount;
		iunseen = gunseen;
		iflagged = gflagged;
		imatched = gmatched;
	} else if (S_ISREG(st.st_mode)) {
		list(0, arg);
	}
}

int
main(int argc, char *argv[])
{
	int c;
	while ((c = getopt(argc, argv, "PRSTDFprstdfX:x:NnCci")) != -1)
		switch (c) {
		case 'P': case 'R': case 'S': case 'T': case 'D': case 'F':
			flags[(unsigned int)c] = 1;
			break;
		case 'p': case 'r': case 's': case 't': case 'd': case 'f':
			flags[(unsigned int)uc(c)] = -1;
			break;
		case 'X':
			while (*optarg)
				flags[(unsigned int)*optarg++] = 1;
			break;
		case 'x':
			while (*optarg)
				flags[(unsigned int)*optarg++] = -1;
			break;
		case 'N': Nflag = 1; break;
		case 'n': Nflag = -1; break;
		case 'C': Cflag = 1; break;
		case 'c': Cflag = -1; break;
		case 'i': iflag = 1; break;
		default:
usage:
			fprintf(stderr,
			    "Usage: mlist [-DFPRST] [-X str]\n"
			    "             [-dfprst] [-x str]\n"
			    "             [-N | -n | -C | -c]\n"
			    "             [-i] [dirs...]\n"
			);
			exit(1);
		}

	int i;

	for (i = 0, flagsum = 0, flagset = 0; (size_t)i < sizeof flags; i++) {
		if (flags[i] != 0)
			flagset++;
		if (flags[i] == 1)
			flagsum++;
	}

	if (optind == argc) {
		if (isatty(0))
			goto usage;
		blaze822_loop(0, 0, listarg);
	} else {
		for (i = optind; i < argc; i++)
			listarg(argv[i]);
	}

	if (iflag && tdirs > 1)
		if (flagset)
			printf("%6ld matched  %6ld unseen  %3ld flagged  %6ld msg\n",
			    tmatched, tunseen, tflagged, tcount);
		else
			printf("%6ld unseen  %3ld flagged  %6ld msg\n",
			    tunseen, tflagged, tcount);

	return 0;
}
